### PR TITLE
Implement ConnectionFactory, and add a container for partitional writ…

### PR DESCRIPTION
**Pull request by vicancy**
…e message

### CODE REVIEW

File: src/Microsoft.Azure.SignalR.AspNet/Infrastructure/ConnectionFactory.cs [hunk 1]
```diff
@@ -40,25 +44,68 @@ public ConnectionFactory(IReadOnlyList<string> hubNames, HubConfiguration hubCon
...
+            foreach (var hub in _hubNames)
+            {
+                // Simply create a couple of connections which connect to Azure SignalR
+                _serviceConnectionManager.AddConnection(hub,
+                    new ServiceConnectionContainer(
+                        () => new ServiceConnection(hub, Guid.NewGuid().ToString(), _protocol, this, _clientConnectionManager, _logger), 
+                        _options.ConnectionCount));
+            }
+
+            Log.StartingConnection(_logger, _name, _options.ConnectionCount, _hubNames.Count + 1);
```
Comment (vwxyzh) [inline src/Microsoft.Azure.SignalR.AspNet/Infrastructure/ConnectionFactory.cs]: why `+1`?

File: src/Microsoft.Azure.SignalR.AspNet/Hubs/ServiceHubDispatcher.cs [hunk 2]
```diff
@@ -39,12 +43,48 @@ public override Task ProcessRequest(HostContext context)
...
+
+            return SendJsonResponse(context, JsonSerializer.Stringify(payload));
         }
 
         private static bool IsNegotiationRequest(IRequest request)
         {
             return request.LocalPath.EndsWith(Constants.Path.Negotiate, StringComparison.OrdinalIgnoreCase);
         }
+
+        private static Task SendJsonResponse(HostContext context, string jsonPayload)
```
Comment (davidfowl) [inline src/Microsoft.Azure.SignalR.AspNet/Hubs/ServiceHubDispatcher.cs]: I'm not sure we support jsonp still.

File: src/Microsoft.Azure.SignalR.AspNet/Infrastructure/ServiceConnectionManager.cs [hunk 3]
```diff
@@ -1,41 +1,60 @@
...
         {
-            throw new System.NotImplementedException();
+            return Task.WhenAll(GetConnections().Select(s => s.WriteAsync(serviceMessage)));
         }
 
-        public Task WriteAsync(string connectionId, object message)
+        private IEnumerable<IServiceConnectionContainer> GetConnections()
         {
-            throw new System.NotImplementedException();
+            foreach(var pair in _serviceConnections)
```
Comment (vwxyzh) [inline src/Microsoft.Azure.SignalR.AspNet/Infrastructure/ServiceConnectionManager.cs]: use linq?

File: src/Microsoft.Azure.SignalR.AspNet/Hubs/ServiceHubDispatcher.cs (hunk 2 above)
Comment (vicancy) [reply src/Microsoft.Azure.SignalR.AspNet/Hubs/ServiceHubDispatcher.cs]: change to throw when jsonp

File: src/Microsoft.Azure.SignalR.AspNet/Infrastructure/ServiceConnectionManager.cs (hunk 3 above)
Comment (vicancy) [reply src/Microsoft.Azure.SignalR.AspNet/Infrastructure/ServiceConnectionManager.cs]: This is a preparation review for an upcoming one, this method is preserved for more changes. 😄

File: src/Microsoft.Azure.SignalR.AspNet/Hubs/ServiceHubDispatcher.cs [hunk 4]
```diff
@@ -39,12 +47,63 @@ public override Task ProcessRequest(HostContext context)
...
+            var advancedClaims = claims?.ToList() ?? new List<Claim>();
+
+            if (authenticationType != null)
+            {
+                advancedClaims.Add(new Claim(Constants.ClaimType.AuthenticationType, authenticationType));
+            }
+
+            advancedClaims.Add(new Claim(Constants.ClaimType.UserId, userId));
+
+            var payload = new
```
Comment (vwxyzh) [inline src/Microsoft.Azure.SignalR.AspNet/Hubs/ServiceHubDispatcher.cs]: remove this anonymous class

File: src/Microsoft.Azure.SignalR.AspNet/Infrastructure/ServiceConnectionManager.cs [hunk 5]
```diff
@@ -1,41 +1,105 @@
...
+                    connections.Add(hub, connection);
+                }
+
+                _serviceConnections = connections;
+            }
+
+            return StartAsync();
         }
 
         public Task StartAsync()
```
Comment (vwxyzh) [inline src/Microsoft.Azure.SignalR.AspNet/Infrastructure/ServiceConnectionManager.cs]: `private`?

File: src/Microsoft.Azure.SignalR.AspNet/Infrastructure/ServiceConnectionContainer.cs [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+        {
+            if (_serviceConnections.Count == 0) return;
+
+            // If we hit this check, it is a code bug.
+            if (string.IsNullOrEmpty(partitionKey))
+            {
+                throw new ArgumentNullException(nameof(partitionKey));
+            }
+
+            var index = Math.Abs(partitionKey.GetHashCode() % _serviceConnections.Count);
```
Comment (vwxyzh) [inline src/Microsoft.Azure.SignalR.AspNet/Infrastructure/ServiceConnectionContainer.cs]: My prefer: `(partitionKey.GetHashCode() & int.MaxValue) % _serviceConnections.Count`

File: src/Microsoft.Azure.SignalR.AspNet/Infrastructure/ServiceConnectionContainer.cs [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+            {
+                throw new ArgumentNullException(nameof(generator));
+            }
+
+            if (count <= 0)
+            {
+                throw new ArgumentException($"{nameof(count)} must be greater than 0.");
+            }
+
+            _serviceConnections = Enumerable.Repeat(generator(), count).ToList();
```
Comment (vwxyzh) [inline src/Microsoft.Azure.SignalR.AspNet/Infrastructure/ServiceConnectionContainer.cs]: Looks array is better.

File: src/Microsoft.Azure.SignalR.AspNet/Infrastructure/ServiceConnectionContainer.cs (hunk 6 above)
Comment (vicancy) [reply src/Microsoft.Azure.SignalR.AspNet/Infrastructure/ServiceConnectionContainer.cs]: cc @xscript How about changing the partition algorithm as we are not yet releasing the SDK?

Comment (xscript) [reply src/Microsoft.Azure.SignalR.AspNet/Infrastructure/ServiceConnectionContainer.cs]: Sure. I am okay with it.

Comment (vicancy) [reply src/Microsoft.Azure.SignalR.AspNet/Infrastructure/ServiceConnectionContainer.cs]: Will update it in another pr

**Pull request merged**